src/libmeasurement_kit/ooni/telegram.cpp

```diff
@@ -0,0 +1,194 @@
+                 cb(NoError());
+             }
+         };
+    };
+
+    for (auto url : urls) {
+        Settings http_options;
+        http_options["http/url"] = url;
+        http::Headers headers = constants::COMMON_CLIENT_HEADERS;
+        std::string body;
+        templates::http_request(entry, http_options, headers, body,
+                                http_cb(url), reactor, logger);
+    }
+
+}
+
+void telegram(std::string input, Settings options,
+              Callback<Var<report::Entry>> callback,
+              Var<Reactor> reactor, Var<Logger> logger) {
+    std::vector<std::string>
+        TELEGRAM_WEB_URLS = { "http://web.telegram.org/",
+                              "https://web.telegram.org/" };
+    // should probably just make these std::pair<std::string,int>,
+    // but I'm not sure if this will be better later when I get
+    // rid of the duplication between this and the http ones
+    std::vector<std::string>
+        TELEGRAM_TCP_ENDPOINTS = { "149.154.175.50:80",
+                                   "149.154.175.50:443",
+                                   "149.154.167.51:80",
+                                   "149.154.167.51:443",
+                                   "149.154.175.100:80",
+                                   "149.154.175.100:443",
+                                   "149.154.167.91:80",
+                                   "149.154.167.91:443",
+                                   "149.154.171.5:80",
+                                   "149.154.171.5:443" };
+
+    // duplication for now to make things easier...
+    std::vector<std::string>
+        TELEGRAM_HTTP_ENDPOINTS = { "http://149.154.175.50:80",
+                                    "http://149.154.175.50:443",
+                                    "http://149.154.167.51:80",
+                                    "http://149.154.167.51:443",
+                                    "http://149.154.175.100:80",
+                                    "http://149.154.175.100:443",
+                                    "http://149.154.167.91:80",
+                                    "http://149.154.167.91:443",
+                                    "http://149.154.171.5:80",
+                                    "http://149.154.171.5:443" };
+
+    logger->info("starting telegram test");
+    Var<Entry> entry(new Entry);
+
+    // first try telegram web
+    http_many(TELEGRAM_WEB_URLS, entry, [=](Error err) {
+        logger->info("done testing telegram web");
+        // then try endpoints as TCP
+        tcp_many(TELEGRAM_TCP_ENDPOINTS, entry, [=](Error err){
+            logger->info("done testing endpoints as TCP");
+            // then try endpoints as HTTP
+            http_many(TELEGRAM_HTTP_ENDPOINTS, entry, [=](Error err){
+                logger->info("done testing endpoints as HTTP");
+                // then finish!
+                callback(entry);
+            }, reactor, logger);
+        }, reactor, logger);
+    }, reactor, logger);
```

Comment:
Speaking of chaining stuff, one thing that I have observed is that, since in most cases we need to pass anyway reactor and logger, it's better to refactor arguments such that the callback is always the last. I have observed that this makes the code more readable than having the callback in the middle and have started to apply this pattern to new code.

Speaking specifically of your general comment regarding chaining functions, this code has the issue that you are ignoring errors, which perhaps is not what we want? I guess we want to use an abstraction that executes a list of functions in sequence (or in parallel) and produces a single error that contains as sub-errors the results of each step; does this make sense to you?